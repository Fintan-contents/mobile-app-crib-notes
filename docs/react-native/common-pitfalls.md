---
title: ハマりがちな落とし穴
---

React Nativeで開発する際に起きた困り事や不具合についてまとめています。
