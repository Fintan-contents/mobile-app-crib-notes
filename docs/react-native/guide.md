---
title: 何らかのガイドを書きたい気持ちはある
---
