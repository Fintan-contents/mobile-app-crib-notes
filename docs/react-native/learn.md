---
title: About
hide_table_of_contents: true
---

このコンテンツはReact Nativeでのモバイルアプリケーション開発を学習するためのコンテンツです。

ReactでWebアプリケーションを実装できるがReact Nativeでモバイルアプリケーションを実装したことがないアプリケーション開発者を対象として、次のことについて学ぶことを目的としています。

- ReactとReact Nativeの違い
- React Nativeで作成したモバイルアプリケーションの動作確認方法

また、作成したモバイルアプリケーションの配布方法やテスト手法についてはこのコンテンツでは触れません。

<!-- textlint-disable ja-technical-writing/sentence-length,ja-technical-writing/max-comma,ja-spacing/ja-no-space-around-parentheses,jtf-style/3.3.かっこ類と隣接する文字の間のスペースの有無,ja-technical-writing/ja-no-mixed-period,ja-technical-writing/no-unmatched-pair -->

import {PageList} from '../../src/components';
const overviews = [
  {
    title: 'Getting Started',
    to: '/react-native/learn/getting-started',
    summary: <>React Nativeでのアプリ開発に必要な環境を用意します。</>
  },
  {
    title: 'アプリの実装前に',
    to: '/react-native/learn/basic-concepts',
    summary: <>React Nativeでアプリを実装し始める前に最低限必要となる前提知識や、React NativeとReact Navigationの基本的な知識について学びます。</>
  },
  {
    title: 'ToDoアプリの実装',
    to: '/react-native/learn/todo-app',
    summary: <>React Nativeで簡単なToDoアプリを実装します。</>
  },
]

<PageList overviews={overviews} colSize={12} />

<!-- textlint-enable ja-technical-writing/sentence-length,ja-technical-writing/max-comma,ja-spacing/ja-no-space-around-parentheses,jtf-style/3.3.かっこ類と隣接する文字の間のスペースの有無,ja-technical-writing/ja-no-mixed-period,ja-technical-writing/no-unmatched-pair -->
