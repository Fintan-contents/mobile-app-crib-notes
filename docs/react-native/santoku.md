---
title: Example Application
sidebar_label: About
hide_table_of_contents: true
---

React Nativeサンプルアプリケーションの設計ドキュメントです。

<!-- textlint-disable ja-technical-writing/sentence-length -->

import {PageList} from '../../src/components';
const overviews = [
]

<PageList overviews={overviews} colSize={12} />

<!-- textlint-enable ja-technical-writing/sentence-length -->
