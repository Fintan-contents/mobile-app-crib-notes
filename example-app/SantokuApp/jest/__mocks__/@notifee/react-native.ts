import notifee from '@notifee/react-native';
export * from '@notifee/react-native/dist/types/Library';
export * from '@notifee/react-native/dist/types/Notification';
export * from '@notifee/react-native/dist/types/Trigger';
export * from '@notifee/react-native/dist/types/NotificationIOS';
export * from '@notifee/react-native/dist/types/NotificationAndroid';

const mock: typeof notifee = {
  cancelAllNotifications: jest.fn(),
  cancelDisplayedNotifications: jest.fn(),
  cancelTriggerNotifications: jest.fn(),
  cancelNotification: jest.fn(),
  cancelDisplayedNotification: jest.fn(),
  cancelTriggerNotification: jest.fn(),
  createChannel: jest.fn(),
  createChannels: jest.fn(),
  createChannelGroup: jest.fn(),
  createChannelGroups: jest.fn(),
  deleteChannel: jest.fn(),
  deleteChannelGroup: jest.fn(),
  displayNotification: jest.fn(),
  openAlarmPermissionSettings: jest.fn(),
  createTriggerNotification: jest.fn(),
  getTriggerNotificationIds: jest.fn(),
  getDisplayedNotifications: jest.fn(),
  getTriggerNotifications: jest.fn(),
  getChannel: jest.fn(),
  isChannelCreated: jest.fn(),
  isChannelBlocked: jest.fn(),
  getChannels: jest.fn(),
  getChannelGroup: jest.fn(),
  getChannelGroups: jest.fn(),
  getInitialNotification: jest.fn(),
  onBackgroundEvent: jest.fn(),
  onForegroundEvent: jest.fn(),
  openNotificationSettings: jest.fn(),
  registerForegroundService: jest.fn(),
  stopForegroundService: jest.fn(),
  requestPermission: jest.fn(),
  setNotificationCategories: jest.fn(),
  getNotificationCategories: jest.fn(),
  getNotificationSettings: jest.fn(),
  getBadgeCount: jest.fn(),
  setBadgeCount: jest.fn(),
  incrementBadgeCount: jest.fn(),
  decrementBadgeCount: jest.fn(),
  openBatteryOptimizationSettings: jest.fn(),
  isBatteryOptimizationEnabled: jest.fn(),
  getPowerManagerInfo: jest.fn(),
  openPowerManagerSettings: jest.fn(),
  hideNotificationDrawer: jest.fn(),
  SDK_VERSION: '5.4.1',
};

// 複数のファイルでnotifeeをimportしていた場合に、redefineされないようにモックが存在するか確認
if (!__mocks.notifee) {
  Object.defineProperty(__mocks, 'notifee', {value: mock});
}

// テストケースごとにモックは初期化しておく。
beforeEach(() =>
  Object.values(mock).forEach(fn => {
    if (jest.isMockFunction(fn)) {
      fn.mockClear();
    }
  }),
);

export default mock;
