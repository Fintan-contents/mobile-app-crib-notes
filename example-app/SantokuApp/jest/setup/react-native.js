// https://reactnavigation.org/docs/testing#mocking-native-modules
// Silence the warning: Animated: `useNativeDriver` is not supported because the native animated module is missing
jest.mock('react-native/Libraries/Animated/NativeAnimatedHelper');

// !!! NOTICE !!!
// The React Navigation documentation says that `react-native-reanimated` and `react-native-gesture-handler`
// also need to be configured with mocks.
// But since React Navigation v6 does not already depend on those packages, mocking those packages is not necessary
// for React Navigation v6.
//
// If you want to use `react-native-reanimated` and `react-native-gesture-handler`, add mock according to the guide
// provided by each package.
// - `react-native-reanimated`: https://docs.swmansion.com/react-native-reanimated/docs/guide/testing/
// - `react-native-gesture-handler`: https://docs.swmansion.com/react-native-gesture-handler/docs/guides/testing
