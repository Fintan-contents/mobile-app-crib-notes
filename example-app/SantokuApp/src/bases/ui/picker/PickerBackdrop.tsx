import {composePressableStyles} from 'bases/core/utils/composePressableStyles';
import React, {useMemo} from 'react';
import {Modal as RNModal, ModalProps, Pressable, PressableProps, StyleSheet, ViewProps} from 'react-native';
import Reanimated, {WithTimingConfig} from 'react-native-reanimated';

import {usePickerBackdrop} from './usePickerBackdrop';

export const DEFAULT_BACKGROUND_COLOR = 'black';
export const DEFAULT_OPACITY = 0.4;
export const DEFAULT_FADE_IN_DURATION = 300;
export const DEFAULT_FADE_OUT_DURATION = 150;

// React Native ReanimatedのLayout Animationsを使用すると↓の不具合が発生するため、指定できないようにしています。
// https://github.com/software-mansion/react-native-reanimated/issues/2906
export type PickerBackdropProps = Omit<Reanimated.AnimateProps<ViewProps>, 'exiting' | 'entering'> & {
  isVisible: boolean;
  onPress?: () => unknown;
  afterFadeIn?: (finished?: boolean) => unknown;
  afterFadeOut?: (finished?: boolean) => unknown;
  fadeInDuration?: number;
  fadeOutDuration?: number;
  opacity?: number;
  pressableProps?: PressableProps;
  modalProps?: ModalProps;
  fadeInConfig?: WithTimingConfig;
  fadeOutConfig?: WithTimingConfig;
};

export const PickerBackdrop: React.FC<React.PropsWithChildren<PickerBackdropProps>> = ({
  isVisible,
  opacity = DEFAULT_OPACITY,
  onPress,
  afterFadeIn,
  afterFadeOut,
  fadeInDuration = DEFAULT_FADE_IN_DURATION,
  fadeOutDuration = DEFAULT_FADE_OUT_DURATION,
  pressableProps: {style: pressableStyle, ...pressableProps} = {},
  modalProps: {style: modalStyle, ...modalProps} = {},
  fadeInConfig,
  fadeOutConfig,
  children,
}) => {
  const {isModalVisible} = usePickerBackdrop({
    isVisible,
    opacity,
    afterFadeIn,
    afterFadeOut,
    fadeInDuration,
    fadeOutDuration,
    fadeInConfig,
    fadeOutConfig,
  });

  const composedPressableStyles = useMemo(
    () => composePressableStyles([styles.pressable, pressableStyle]),
    [pressableStyle],
  );

  return !isModalVisible ? null : (
    <RNModal
      visible
      statusBarTranslucent
      animationType="fade"
      transparent
      // 戻るボタンが押されたとき（onRequestClose）は、背景がタップされたときと同じ振る舞いになるようにしておく。
      onRequestClose={onPress}
      style={modalStyle}
      {...modalProps}>
      <Pressable onPress={onPress} style={composedPressableStyles} {...pressableProps} />
      {children}
    </RNModal>
  );
};

const styles = StyleSheet.create({
  pressable: {
    ...StyleSheet.absoluteFillObject,
    display: 'flex',
    opacity: 0.4,
    backgroundColor: DEFAULT_BACKGROUND_COLOR,
  },
  backdrop: {
    flex: 1,
  },
});
