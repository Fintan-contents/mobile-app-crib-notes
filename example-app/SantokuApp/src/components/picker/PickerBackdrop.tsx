import {composePressableStyles} from 'framework/utilities';
import React, {useMemo} from 'react';
import {Modal as RNModal, ModalProps, Omit, Pressable, PressableProps, StyleSheet, ViewProps} from 'react-native';
import Reanimated, {BaseAnimationBuilder, FadeIn, FadeOut, Keyframe} from 'react-native-reanimated';

import {usePickerBackdropUseCase} from './usePickerBackdropUseCase';

export const DEFAULT_COLOR = 'rgba(0,0,0,0.4)';
export const DEFAULT_FADE_IN_DURATION = 300;
export const DEFAULT_FADE_OUT_DURATION = 150;
export const DEFAULT_ENTERING = FadeIn.duration(DEFAULT_FADE_IN_DURATION);
export const DEFAULT_EXITING = FadeOut.duration(DEFAULT_FADE_OUT_DURATION);

export type PickerBackdropProps = Omit<Reanimated.AnimateProps<ViewProps>, 'entering' | 'exiting'> & {
  isVisible: boolean;
  onPress?: () => unknown;
  /**
   * iOSの場合、アニメーションが終わった後に呼び出されます。
   * Androidの場合、アニメーションが始まった時に呼び出されます。
   */
  enteringCallback?: (finished: boolean) => unknown;
  exitingCallback?: (finished: boolean) => unknown;
  pressableProps?: PressableProps;
  modalProps?: ModalProps;
  /**
   * enteringに指定したAnimationBuilderなどでwithCallbackを指定しても、このコンポーネントの中で上書きしているため実行できません。
   * withCallbackで実行する関数は、enteringCallbackで指定してください。
   */
  entering?: BaseAnimationBuilder | Keyframe;
  /**
   * exitingに指定したAnimationBuilderなどでwithCallbackを指定しても、このコンポーネントの中で上書きしているため実行できません。
   * withCallbackで実行する関数は、exitingCallbackで指定してください。
   */
  exiting?: BaseAnimationBuilder | Keyframe;
};

export const PickerBackdrop: React.FC<PickerBackdropProps> = ({
  isVisible,
  onPress,
  entering = DEFAULT_ENTERING,
  exiting = DEFAULT_EXITING,
  enteringCallback,
  exitingCallback,
  pressableProps: {style: pressableStyle, ...pressableProps} = {},
  modalProps: {style: modalStyle, ...modalProps} = {},
  /**
   * このコンポーネントでは、ReanimatedのEntering/Exitingを使用してアニメーションを実現しています。
   * Entering/Exitingを使用した場合は、opacityを設定しても反映されません。
   * backgroundColorにrgbaで指定してください。（e.g. backgroundColor: rgba(0,0,0,0.4)）
   */
  style,
  children,
  ...animatedViewProps
}) => {
  const {isModalVisible, composedEnteringCallback, composedExitingCallback} = usePickerBackdropUseCase({
    isVisible,
    enteringCallback,
    exitingCallback,
  });

  const composedPressableStyles = useMemo(
    () => composePressableStyles([styles.pressable, pressableStyle]),
    [pressableStyle],
  );

  return !isModalVisible ? null : (
    <RNModal
      visible
      statusBarTranslucent
      animationType="none"
      transparent
      // 戻るボタンが押されたとき（onRequestClose）は、背景がタップされたときと同じ振る舞いになるようにしておく。
      onRequestClose={onPress}
      style={modalStyle}
      {...modalProps}>
      {isVisible && (
        <Pressable onPress={onPress} style={composedPressableStyles} {...pressableProps}>
          <Reanimated.View
            entering={entering.withCallback(composedEnteringCallback)}
            exiting={exiting.withCallback(composedExitingCallback)}
            style={StyleSheet.flatten([styles.backdrop, style])}
            {...animatedViewProps}
          />
        </Pressable>
      )}
      {children}
    </RNModal>
  );
};

const styles = StyleSheet.create({
  pressable: {
    ...StyleSheet.absoluteFillObject,
    display: 'flex',
  },
  backdrop: {
    flex: 1,
    backgroundColor: DEFAULT_COLOR,
  },
});
