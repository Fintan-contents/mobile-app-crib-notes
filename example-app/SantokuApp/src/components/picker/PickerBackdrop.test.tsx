import {fireEvent, render} from '@testing-library/react-native';
import React from 'react';
import {ModalProps, PressableProps, ViewProps} from 'react-native';
import Reanimated, {ZoomIn, ZoomOut} from 'react-native-reanimated';

import {PICKER_BACKDROP_DEFAULT_ENTERING, PICKER_BACKDROP_DEFAULT_EXITING, PickerBackdrop} from './PickerBackdrop';

// If advancing a timer changes the state of a component, the timer must be run within an act.
// However, since act is `Thenable`, ESLint will issue a warning if you do not do something like await.
// For convenience, disable the relevant rule in this file.
/* eslint-disable @typescript-eslint/no-floating-promises */

// To get react-native-reanimated v2 animations to work with timer mocks, you need to use the `modern` timer.
// By default, Jest v26 uses `legacy` timer. So if you use `withReanimatedTimer` that use the default timer, animation does not start.
// https://jestjs.io/ja/docs/timer-mocks
jest.useFakeTimers('modern');

// TODO: Jest v27にアップデートできたら、withReanimatedTimerでテストを実装できるか検証する。
//       （JestのバージョンはExpoに依存しているので、Expoでのアップデートを待っている状態）

jest.runAllTimers();

describe('PickerBackdrop only with required props', () => {
  it('returns null if not visible', () => {
    const sut = render(<PickerBackdrop isVisible={false} />);
    // PickerBackdropがnullを返していることを確認したいがうまくやる方法が見当たらないので`toJSON`でnullになることを確認する。
    expect(sut.toJSON()).toBeNull();
  });

  it('renders successfully only with required props', () => {
    const sut = render(<PickerBackdrop isVisible testID="backdropAnimated" />);
    const animatedView = sut.getByTestId('backdropAnimated');
    const animatedViewProps = animatedView.props as Reanimated.AnimateProps<ViewProps>;
    // Animated.Viewのentering/exitingをテストで実行することができなかったため、entering/exitingにデフォルトアニメーションが設定されていることのみを確認する。
    expect(sut).toMatchSnapshot('AnimatedView with visible.');
    expect(animatedView).not.toBeNull();
    expect(animatedViewProps.entering).toBe(PICKER_BACKDROP_DEFAULT_ENTERING);
    expect(animatedViewProps.exiting).toBe(PICKER_BACKDROP_DEFAULT_EXITING);

    //////////////////////////////////////////////////////////////////////////////////
    // 非表示にする
    //////////////////////////////////////////////////////////////////////////////////
    sut.update(<PickerBackdrop isVisible={false} />);
    const animatedView2 = sut.queryByTestId('backdropAnimated');
    expect(sut).toMatchSnapshot('AnimatedView with invisible.');
    expect(animatedView2).toBeNull();
  });
});

describe('PickerBackdrop with `onPress', () => {
  it('should be called on pressed', () => {
    const onPress = jest.fn();
    const sut = render(
      <PickerBackdrop
        isVisible
        onPress={onPress}
        testID="pickerBackDrop"
        pressableProps={{testID: 'pressable'}}
        modalProps={{testID: 'modal'}}
      />,
    );
    fireEvent.press(sut.getByTestId('pressable'));
    expect(onPress).toHaveBeenCalledTimes(1);
    fireEvent(sut.getByTestId('modal'), 'onRequestClose');
    expect(onPress).toHaveBeenCalledTimes(2);
  });
});
//
describe('PickerBackdrop with all props', () => {
  it('should be applied properly', () => {
    const onPress = jest.fn();
    const onLongPress = jest.fn();
    const entering = ZoomIn.duration(500);
    const exiting = ZoomOut.duration(300);
    /**
     * entering/exitingを実行できなかったため、以下のPropsは検証できていません
     * - enteringCallback
     * - exitingCallback
     *
     * animatedPropsは取得できなかったため（Snapshot上にも存在していない）、検証できていません
     */
    const sut = render(
      <PickerBackdrop
        isVisible
        testID="animatedView"
        onPress={onPress}
        modalProps={{
          testID: 'modal',
          style: {backgroundColor: 'yellow'},
          statusBarTranslucent: false,
          animationType: 'fade',
          transparent: false,
          presentationStyle: 'fullScreen',
        }}
        pressableProps={{testID: 'pressable', style: {display: 'none'}, onLongPress}}
        animatedProps={{pointerEvents: 'none'}}
        style={{backgroundColor: 'green', borderColor: 'red'}}
        entering={entering}
        exiting={exiting}
      />,
    );
    expect(sut).toMatchSnapshot('PickerBackdrop with all props.');
    const modal = sut.getByTestId('modal');
    const pressable = sut.getByTestId('pressable');
    const animatedView = sut.getByTestId('animatedView');

    // assert modal
    fireEvent(modal, 'onRequestClose');
    const modalProps = modal.props as ModalProps;
    expect(modalProps.visible).toBe(true);
    expect(modalProps.statusBarTranslucent).toBe(false);
    expect(modalProps.animationType).toBe('fade');
    expect(modalProps.transparent).toBe(false);
    expect(modalProps.style).toEqual({backgroundColor: 'yellow'});
    expect(modalProps.presentationStyle).toBe('fullScreen');
    expect(onPress).toHaveBeenCalledTimes(1);

    // assert pressable
    fireEvent.press(pressable);
    fireEvent(pressable, 'onLongPress');
    const pressableProps = pressable.props as PressableProps;
    expect(pressableProps.style).toEqual({bottom: 0, display: 'none', left: 0, position: 'absolute', right: 0, top: 0});
    expect(onPress).toHaveBeenCalledTimes(2);
    expect(onLongPress).toHaveBeenCalledTimes(1);

    // assert animatedView
    const animatedViewProps = animatedView.props as Reanimated.AnimateProps<ViewProps>;
    expect(animatedViewProps.style).toEqual({flex: 1, backgroundColor: 'green', borderColor: 'red'});
    expect(animatedViewProps.entering).toBe(entering);
    expect(animatedViewProps.exiting).toBe(exiting);
  });
});
