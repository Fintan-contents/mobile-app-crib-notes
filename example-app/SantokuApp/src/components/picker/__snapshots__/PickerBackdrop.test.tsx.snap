// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PickerBackdrop only with required props renders successfully only with required props: AnimatedView with invisible. 1`] = `
<Modal
  animationType="none"
  hardwareAccelerated={false}
  statusBarTranslucent={true}
  transparent={true}
  visible={true}
/>
`;

exports[`PickerBackdrop only with required props renders successfully only with required props: AnimatedView with visible. 1`] = `
<Modal
  animationType="none"
  hardwareAccelerated={false}
  statusBarTranslucent={true}
  transparent={true}
  visible={true}
>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "bottom": 0,
        "display": "flex",
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <View
      animatedStyle={
        Object {
          "value": Object {},
        }
      }
      collapsable={false}
      entering={
        FadeIn {
          "build": [Function],
          "callbackV": [Function],
          "durationV": 300,
          "randomizeDelay": false,
        }
      }
      exiting={
        FadeOut {
          "build": [Function],
          "callbackV": [Function],
          "durationV": 150,
          "randomizeDelay": false,
        }
      }
      style={
        Object {
          "backgroundColor": "rgba(0,0,0,0.4)",
          "flex": 1,
        }
      }
      testID="backdropAnimated"
    />
  </View>
</Modal>
`;

exports[`PickerBackdrop with all props should be applied properly: PickerBackdrop with all props. 1`] = `
<Modal
  animationType="fade"
  hardwareAccelerated={false}
  onRequestClose={[MockFunction]}
  presentationStyle="fullScreen"
  statusBarTranslucent={false}
  style={
    Object {
      "backgroundColor": "yellow",
    }
  }
  testID="modal"
  transparent={false}
  visible={true}
>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "bottom": 0,
        "display": "none",
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
    testID="pressable"
  >
    <View
      animatedStyle={
        Object {
          "value": Object {},
        }
      }
      collapsable={false}
      entering={
        ZoomIn {
          "build": [Function],
          "callbackV": [Function],
          "durationV": 500,
          "randomizeDelay": false,
        }
      }
      exiting={
        ZoomOut {
          "build": [Function],
          "callbackV": [Function],
          "durationV": 300,
          "randomizeDelay": false,
        }
      }
      style={
        Object {
          "backgroundColor": "green",
          "borderColor": "red",
          "flex": 1,
        }
      }
      testID="animatedView"
    />
  </View>
</Modal>
`;
